[SYCL][Docs] Improve Docs (#99)
triSYCL/sycl: sycl/doc/GettingStartedXilinxFPGA.md

## Compile the SYCL compiler

> building sycl can be done with Python scripts
> environment:

Reviewer: ```suggestion
Building SYCL can be done with Python scripts:
```
Ahem...

> environment:

Reviewer: I guess this line is a left-over

in the second log we can find the specific `v++` command that has an issue.

#### llvm-reduce

It is possible to use `llvm-reduce` to track down `v++` issues.
First build `llvm-reduce` with:
```bash
ninja -C build-Release llvm-reduce
```
then build a script called later `is_interesting_llvm.sh` to exhibits the `v++` bug.
This will look like the following:
```bash
#!/bin/bash

# Downgrade the IR generate by llvm-reduce
./build-Release/bin/opt -verify -S -xoccIRDowngrader $1 -o $1.tmp.ll || exit 1

# Assemble the IR using Vitis's assembler
.../clang-3.9-csynth/bin/llvm-as $1.tmp.ll -o $1.tmp.xpirbc

# run the command that crashed v++ on the file generated by llvm-reduce while outputting everything to a file
.../clang-3.9-csynth/bin/clang ... -x ir $1.tmp.xpirbc -o - &> out

# Save the exit code of the command since it might have crashed
res=$?

# Test that the output matches the original error
cat out | grep "..." > /dev/null || exit 1

# Test the error code
if [ $res -eq 254 ]; then
# exit 0 to indicate this still exhibits the original bug
  exit 0
else

> # exit 1 to indicate this doesn't exhibits the original bug

Reviewer: ```suggestion
# exit 1 to indicate this doesn't exhibit the original bug
```

 -DCMAKE_EXPORT_COMPILE_COMMANDS=1 \
 -G Ninja \
 -DLLVM_ENABLE_ASSERTIONS=ON \
 -DCMAKE_C_COMPILER="/usr/bin/clang-11" \
 -DCMAKE_CXX_COMPILER="/usr/bin/clang++-11" \
 -DLLVM_USE_LINKER="lld-11" \
 -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" \
 -DLLVM_EXTERNAL_PROJECTS="sycl;llvm-spirv;opencl-aot;xpti;libdevice" \
 -DLLVM_EXTERNAL_SYCL_SOURCE_DIR=$SYCL_HOME/sycl \
 -DLLVM_EXTERNAL_LLVM_SPIRV_SOURCE_DIR=$SYCL_HOME/llvm-spirv \
 -DLLVM_EXTERNAL_XPTI_SOURCE_DIR=$SYCL_HOME/xpti \
 -DLLVM_EXTERNAL_LIBDEVICE_SOURCE_DIR=$SYCL_HOME/libdevice \
 -DLLVM_ENABLE_PROJECTS="clang;llvm-spirv;sycl;opencl-aot;xpti;libdevice;libclc" \
 -DLIBCLC_TARGETS_TO_BUILD="nvptx64--;nvptx64--nvidiacl" \
 -DSYCL_BUILD_PI_CUDA=ON \
 -DLLVM_BUILD_TOOLS=ON \
 -DSYCL_INCLUDE_TESTS=ON \
 -DLLVM_ENABLE_DOXYGEN=OFF \
 -DLLVM_ENABLE_SPHINX=OFF \
 -DBUILD_SHARED_LIBS=ON \
 -DSYCL_ENABLE_XPTI_TRACING=ON \
 $SYCL_HOME/llvm
```
then build with
```bash
cd $SYCL_HOME
ninja -C build-Debug sycl-toolchain
```

* `-CMAKE_CXX_FLAGS_RELEASE="-O3 -march=native -fno-omit-frame-pointer"` to have an optimized build with relatively accurate stack traces;
* `-G Ninja` use `ninja` instead of `make` to speedup builds;

> * `-DLLVM_USE_LINKER="lld-11"` select `lld` or `gold` instead of `ld` to speedup builds

Reviewer: ```suggestion
* `-DLLVM_USE_LINKER="lld-11"` select `lld` or `gold` instead of `ld` to speedup builds;
```

but the more detailed log file `ROOT/_x/sycl-xocc.ppArYN_kernel_name/kernel_name/kernel_name/solution/.autopilot/db/autopilot.flow.log`

Then in this log we can find the specific command `v++` was running when it crashed.

#### llvm-reduce

It is possible to use `llvm-reduce` to track down `v++` issues.
First build `llvm-reduce` with:
```bash
ninja -C build-Release llvm-reduce
```
then build a script called later `is_interesting_llvm.sh` to exhibit the `v++` bug.
This will look like the following:
```bash
#!/bin/bash

# Downgrade the IR generate by llvm-reduce
./build-Release/bin/opt -verify -S -xoccIRDowngrader $1 -o $1.tmp.ll || exit 1

# Assemble the IR using Vitis's assembler
.../clang-3.9-csynth/bin/llvm-as $1.tmp.ll -o $1.tmp.xpirbc

# run the command that crashed v++ on the file generated by llvm-reduce while outputting everything to a file
.../clang-3.9-csynth/bin/clang ... -x ir $1.tmp.xpirbc -o - &> out

# Save the exit code of the command since it might have crashed
res=$?

# Test that the output matches the original error
cat out | grep "..." > /dev/null || exit 1

# Test the error code
if [ $res -eq 254 ]; then

> # exit 0 to indicate this still exhibit the original bug

Reviewer: ```suggestion
# exit 0 to indicate this still exhibits the original bug
```

 -DCMAKE_INSTALL_PREFIX="../Install-Debug" \
 -DCMAKE_EXPORT_COMPILE_COMMANDS=1 \
 -G Ninja \
 -DLLVM_ENABLE_ASSERTIONS=ON \
 -DCMAKE_C_COMPILER="/usr/bin/clang-11" \
 -DCMAKE_CXX_COMPILER="/usr/bin/clang++-11" \
 -DLLVM_USE_LINKER="lld-11" \
 -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" \
 -DLLVM_EXTERNAL_PROJECTS="sycl;llvm-spirv;opencl-aot;xpti;libdevice" \
 -DLLVM_EXTERNAL_SYCL_SOURCE_DIR=$SYCL_HOME/sycl \
 -DLLVM_EXTERNAL_LLVM_SPIRV_SOURCE_DIR=$SYCL_HOME/llvm-spirv \
 -DLLVM_EXTERNAL_XPTI_SOURCE_DIR=$SYCL_HOME/xpti \
 -DLLVM_EXTERNAL_LIBDEVICE_SOURCE_DIR=$SYCL_HOME/libdevice \
 -DLLVM_ENABLE_PROJECTS="clang;llvm-spirv;sycl;opencl-aot;xpti;libdevice;libclc" \
 -DLIBCLC_TARGETS_TO_BUILD="nvptx64--;nvptx64--nvidiacl" \
 -DSYCL_BUILD_PI_CUDA=ON \
 -DLLVM_BUILD_TOOLS=ON \
 -DSYCL_INCLUDE_TESTS=ON \
 -DLLVM_ENABLE_DOXYGEN=OFF \
 -DLLVM_ENABLE_SPHINX=OFF \
 -DBUILD_SHARED_LIBS=ON \
 -DSYCL_ENABLE_XPTI_TRACING=ON \
 $SYCL_HOME/llvm
```
then build with
```bash
cd $SYCL_HOME
ninja -C build-Debug sycl-toolchain
```

* `-CMAKE_CXX_FLAGS_RELEASE="-O3 -march=native -fno-omit-frame-pointer"` to have an optimized build with relatively accurate stack traces;

> * `-G Ninja` use ninja instead of make to speedup builds;

Reviewer: ```suggestion
* `-G Ninja` use `ninja` instead of `make` to speedup builds;
```

* `-DLLVM_USE_LINKER="lld-11"` select `lld` or `gold` instead of `ld` to speedup builds

> * `-DLLVM_TARGETS_TO_BUILD="X86;NVPTX"` to only build needed targets NVPTX is only needed for CUDA support;

Reviewer: ```suggestion
* `-DLLVM_TARGETS_TO_BUILD="X86;NVPTX"` to build only the targets needed, like NVPTX for CUDA support;
```

* `-DLIBCLC_TARGETS_TO_BUILD="nvptx64--;nvptx64--nvidiacl"` needed for CUDA support;
* `-DSYCL_ENABLE_XPTI_TRACING=ON` adds useful debugging capabilities;

> * `-DBUILD_SHARED_LIBS=ON` to build speedup;

Reviewer: ```suggestion
* `-DBUILD_SHARED_LIBS=ON` to speed up the build process by using shared libraries;
```

> * `-DLLVM_TABLEGEN="$SYCL_HOME/build-Release/bin/llvm-tblgen" -DCLANG_TABLEGEN="$SYCL_HOME/build-Release/bin/clang-tblgen"` to reuse the tablegen built with the release version and speedup builds. The release build must be done before configuring the Debug one;

Reviewer: ```suggestion
* `-DLLVM_TABLEGEN="$SYCL_HOME/build-Release/bin/llvm-tblgen" -DCLANG_TABLEGEN="$SYCL_HOME/build-Release/bin/clang-tblgen"` to reuse the tablegen output built with the Release version and speedup builds. The Release build must be done before configuring the Debug one;
```

Reviewer: I still think that Release and Debug in that case need a capital letter since it refers to the LLVM model.

* `-DCMAKE_CXX_FLAGS_DEBUG="-g -fstandalone-debug"` to have full debug information.

> * For details about the CMake configuration see https://llvm.org/docs/CMake.html

Reviewer: ```suggestion
For details about the `CMake` configuration see https://llvm.org/docs/CMake.html
```

> while building the sycl-toolchain target, the device runtime for spirv and nvptx targets gets compiled using the device compiler from the specific build.

Reviewer: ```suggestion
While building the `sycl-toolchain` target, the device runtime for `spirv` and `nvptx` targets gets compiled using the device compiler from the specific build.
```

> building anything with a debug compiler is very slow, but it can be speedup using:
> 

Reviewer: ```suggestion
Building anything with a debug compiler is very slow, but it can be speedup using:
```

```bash
export LD_LIBRARY_PATH=$SYCL_HOME/build-Release/lib:$LD_LIBRARY_PATH
```

> 
> this will make the debug compiler select the dynamic libraries of the release compiler and speedup the build. This only works because the Release and Debug build have ABI compatible configuration, changing LLVM_ENABLE_ASSERTIONS or other configuration may change that.

Reviewer: ```suggestion

This will make the debug compiler select the dynamic libraries of the release compiler and speedup the build. This only works because the Release and Debug build have ABI compatible configuration, changing `LLVM_ENABLE_ASSERTIONS` or other configuration may change that.
```

### Debugging

#### Environnement variables

> some environment variables are very useful for debugging:

Reviewer: ```suggestion
Some environment variables are very useful for debugging:
```

```bash

> # Redirect the directory used as temporary directory 

Reviewer: ```suggestion
# Redirect the directory used as temporary directory
```

# for the compiler and various tools
export TMP=$SYCL_HOME/../tmp

# SYCL_PI_TRACE should always be at least at 1, this make the SYCL runtime emit logs about which device is selected
export SYCL_PI_TRACE=1

# SYCL_PI_TRACE can be set at -1 to have full debug information but this is quite noisy
export SYCL_PI_TRACE=-1
```

#### Clang flags

Some useful Clang flags:
* `-ccc-print-phases` outputs the graph of compilation phases;

> * `-ccc-print-bindings` outputs the mapping from compilation phases to commands.

Reviewer: ```suggestion
* `-ccc-print-bindings` outputs the mapping from compilation phases to commands;
```

* `-###` outputs the commands to run to compile. The driver will create files to reserve them for those commands.

#### Running a single test

> To run test of the test suite in isolation use:

Reviewer: ```suggestion
To run a test from the test suite in isolation, use:
```

```bash
python3 /path/to/build/dir/bin/llvm-lit -v --param XOCC=all path/to/test.cpp
```
where all tests utilities must have been build for this to work.

#### v++ Logs
The kinds of following errors are typical of a back-end issue:
```
ERROR: [v++ 60-300] Failed to build kernel(ip) kernel_name, see log for details: ROOT/_x/sycl-xocc.ppArYN_kernel_name/kernel_name/vitis_hls.log
ERROR: [v++ 60-599] Kernel compilation failed to complete
ERROR: [v++ 60-592] Failed to finish compilation
```

> the path contains a hash and kernel names so they wont match exactly.

Reviewer: ```suggestion
the path contains a hash and kernel names so they won't match exactly.
```

Sometimes the specified log file contains useful information

> but the more detailed log file `ROOT/_x/sycl-xocc.ppArYN_kernel_name/kernel_name/kernel_name/solution/.autopilot/db/autopilot.flow.log`

Reviewer: Fix the cohesion of the sentence. I do not even if you are talking about 1 or 2 different log files here...

Reviewer: Then?

Then in this log we can find the specific command `v++` was running when it crashed.

#### llvm-reduce

It is possible to use `llvm-reduce` to track down `v++` issues.
First build `llvm-reduce` with:
```bash
ninja -C build-Release llvm-reduce
```
then build a script called later `is_interesting_llvm.sh` to exhibit the `v++` bug.
This will look like the following:
```bash
#!/bin/bash

# Downgrade the IR generate by llvm-reduce
./build-Release/bin/opt -verify -S -xoccIRDowngrader $1 -o $1.tmp.ll || exit 1

# Assemble the IR using Vitis's assembler
.../clang-3.9-csynth/bin/llvm-as $1.tmp.ll -o $1.tmp.xpirbc

> # run the command that crashed v++ on the file generated by llvm-reduce while outputting everything to a file.

Reviewer: ```suggestion
# run the command that crashed v++ on the file generated by llvm-reduce while outputting everything to a file
```

.../clang-3.9-csynth/bin/clang ... -x ir $1.tmp.xpirbc -o - &> out

# Save the exit code of the command since it might have crashed
res=$?

# Test that the output matches the original error
cat out | grep "..." > /dev/null || exit 1

# Test the error code
if [ $res -eq 254 ]; then
# exit 0 to indicate this still exhibit the original bug
  exit 0
else
# exit 1 to indicate this doesn't exhibit the original bug
  exit 1
fi
```

> than we can run llvm-reduce by doing the following:

Reviewer: ```suggestion
then we can run `llvm-reduce` by doing the following:
```

```bash
# Disassemble the file that crashed Vitis's clang
./build-Release/bin/opt -S file_that_crashed_vitis_clang.bc -o tmp.ll

# Run llvm-reduce
./build-Release/bin/llvm-reduce --test=./is_interesting_llvm.sh tmp.ll
```

> When llvm-reduce finishes, the reduced IR crashing `v++` will be in a file named `reduced.ll`.

Reviewer: ```suggestion
When `llvm-reduce` finishes, the reduced IR crashing `v++` will be in a file named `reduced.ll`.
```

  assumptions are false. However, in the basic 2018.3 release the
  standard directory structure that is assumed is correct without
  alterations.

> ## Debugging SYCL

Reviewer: ```suggestion
## Debugging the SYCL implementation
```

### Build

> for serious work on the sycl toolchain it is better to

Reviewer: ```suggestion
For serious work on the SYCL toolchain it is better to
```
For serious work it is better to spell correctly the project we are working on ;-)

> not use the scripts in $SYCL_HOME/buildbot but to invoke cmake directly.

Reviewer: ```suggestion
not use the scripts in `$SYCL_HOME/buildbot` but to invoke `cmake` directly.
```

> this gives much more control over the build configuration

Reviewer: ```suggestion
This gives much more control over the build configuration.
```

> personally I work with 2 builds a Release on used by default

Reviewer: ```suggestion
It is quite useful to work with two builds, a Release one used by default
```

and a Debug one used for debugging.

> Note: the configuration of environnement variables must be done before the cmake invocation.

Reviewer: ```suggestion
Note: the configuration of environment variables must be done before the `cmake` invocation.
```

> here is my Release configuration and build script

Reviewer: ```suggestion
A possible Release configuration and build script targeting Xilinx FPGA, CUDA & OpenCL:
```

cd $SYCL_HOME
mkdir -p "build-Release" && cd "build-Release" && cmake \
 -DCMAKE_INSTALL_PREFIX="../Install-Release" \
 -DCMAKE_CXX_FLAGS_RELEASE="-O3 -march=native -fno-omit-frame-pointer" \
 -DCMAKE_EXPORT_COMPILE_COMMANDS=1 \
 -G Ninja \
 -DCMAKE_C_COMPILER="/usr/bin/clang-11" \
 -DCMAKE_CXX_COMPILER="/usr/bin/clang++-11" \
 -DLLVM_USE_LINKER="lld-11" \
 -DCMAKE_BUILD_TYPE=Release \
 -DLLVM_ENABLE_ASSERTIONS=ON \
 -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" \
 -DLLVM_EXTERNAL_PROJECTS="sycl;llvm-spirv;opencl-aot;xpti;libdevice" \
 -DLLVM_EXTERNAL_SYCL_SOURCE_DIR=$SYCL_HOME/sycl \
 -DLLVM_EXTERNAL_LLVM_SPIRV_SOURCE_DIR=$SYCL_HOME/llvm-spirv \
 -DLLVM_EXTERNAL_XPTI_SOURCE_DIR=$SYCL_HOME/xpti \
 -DLLVM_EXTERNAL_LIBDEVICE_SOURCE_DIR=$SYCL_HOME/libdevice \
 -DLLVM_ENABLE_PROJECTS="clang;llvm-spirv;sycl;opencl-aot;xpti;libdevice;libclc" \
 -DLIBCLC_TARGETS_TO_BUILD="nvptx64--;nvptx64--nvidiacl" \
 -DSYCL_BUILD_PI_CUDA=ON \
 -DLLVM_BUILD_TOOLS=ON \
 -DSYCL_INCLUDE_TESTS=ON \
 -DLLVM_ENABLE_DOXYGEN=OFF \
 -DLLVM_ENABLE_SPHINX=OFF \
 -DBUILD_SHARED_LIBS=ON \
 -DSYCL_ENABLE_XPTI_TRACING=ON \
 $SYCL_HOME/llvm
```
then build with
```bash
cd $SYCL_HOME
ninja -C build-Release sycl-toolchain
```

> My Debug configuration script

Reviewer: ```suggestion
A possible Debug configuration and build script targeting Xilinx FPGA, CUDA & OpenCL:
```

 -DCMAKE_CXX_FLAGS_DEBUG="-g -fstandalone-debug" \
 -DLLVM_TABLEGEN="$(pwd)/../build-Release/bin/llvm-tblgen" \
 -DCLANG_TABLEGEN="$(pwd)/../build-Release/bin/clang-tblgen" \
 -DCMAKE_INSTALL_PREFIX="../Install-Debug" \
 -DCMAKE_EXPORT_COMPILE_COMMANDS=1 \
 -G Ninja \
 -DLLVM_ENABLE_ASSERTIONS=ON \
 -DCMAKE_C_COMPILER="/usr/bin/clang-11" \
 -DCMAKE_CXX_COMPILER="/usr/bin/clang++-11" \
 -DLLVM_USE_LINKER="lld-11" \
 -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" \
 -DLLVM_EXTERNAL_PROJECTS="sycl;llvm-spirv;opencl-aot;xpti;libdevice" \
 -DLLVM_EXTERNAL_SYCL_SOURCE_DIR=$SYCL_HOME/sycl \
 -DLLVM_EXTERNAL_LLVM_SPIRV_SOURCE_DIR=$SYCL_HOME/llvm-spirv \
 -DLLVM_EXTERNAL_XPTI_SOURCE_DIR=$SYCL_HOME/xpti \
 -DLLVM_EXTERNAL_LIBDEVICE_SOURCE_DIR=$SYCL_HOME/libdevice \
 -DLLVM_ENABLE_PROJECTS="clang;llvm-spirv;sycl;opencl-aot;xpti;libdevice;libclc" \
 -DLIBCLC_TARGETS_TO_BUILD="nvptx64--;nvptx64--nvidiacl" \
 -DSYCL_BUILD_PI_CUDA=ON \
 -DLLVM_BUILD_TOOLS=ON \
 -DSYCL_INCLUDE_TESTS=ON \
 -DLLVM_ENABLE_DOXYGEN=OFF \
 -DLLVM_ENABLE_SPHINX=OFF \
 -DBUILD_SHARED_LIBS=ON \
 -DSYCL_ENABLE_XPTI_TRACING=ON \
 $SYCL_HOME/llvm
```
then build with
```bash
cd $SYCL_HOME
ninja -C build-Debug sycl-toolchain
```

> * ```-CMAKE_CXX_FLAGS_RELEASE="-O3 -march=native -fno-omit-frame-pointer"``` to have an optimized build with relatively accurate stack traces

Reviewer: ```suggestion
* `-CMAKE_CXX_FLAGS_RELEASE="-O3 -march=native -fno-omit-frame-pointer"` to have an optimized build with relatively accurate stack traces;
```

> * ```-G Ninja``` use ninja instead of make to speedup builds

Reviewer: ```suggestion
* `-G Ninja` use ninja instead of make to speedup builds;
```

> * ```-DLLVM_USE_LINKER="lld-11"``` select lld or gold instead of ld to speedup builds

Reviewer: ```suggestion
* `-DLLVM_USE_LINKER="lld-11"` select `lld` or `gold` instead of `ld` to speedup builds
```

> * ```-DLLVM_TARGETS_TO_BUILD="X86;NVPTX"``` to only build needed targets NVPTX is only needed for CUDA support.

Reviewer: ```suggestion
* `-DLLVM_TARGETS_TO_BUILD="X86;NVPTX"` to only build needed targets NVPTX is only needed for CUDA support;
```

> * ```-DLIBCLC_TARGETS_TO_BUILD="nvptx64--;nvptx64--nvidiacl"``` needed for cuda support

Reviewer: ```suggestion
* `-DLIBCLC_TARGETS_TO_BUILD="nvptx64--;nvptx64--nvidiacl"` needed for CUDA support;
```

> * ```-DSYCL_ENABLE_XPTI_TRACING=ON``` adds useful debugging capabilities.

Reviewer: ```suggestion
* `-DSYCL_ENABLE_XPTI_TRACING=ON` adds useful debugging capabilities;
```

> * ```-DBUILD_SHARED_LIBS=ON``` to build speedup.

Reviewer: ```suggestion
* `-DBUILD_SHARED_LIBS=ON` to build speedup;
```

> * ```-DLLVM_TABLEGEN="$SYCL_HOME/build-Release/bin/llvm-tblgen" -DCLANG_TABLEGEN="$SYCL_HOME/build-Release/bin/clang-tblgen" ``` to reuse the tablgen built with the release version and speedup builds. the release build must be done before configuring the Debug one.

Reviewer: ```suggestion
* `-DLLVM_TABLEGEN="$SYCL_HOME/build-Release/bin/llvm-tblgen" -DCLANG_TABLEGEN="$SYCL_HOME/build-Release/bin/clang-tblgen"` to reuse the tablegen built with the release version and speedup builds. The release build must be done before configuring the Debug one;
```

> * ```-DCMAKE_CXX_FLAGS_DEBUG="-g -fstandalone-debug"``` to have full debug information.

Reviewer: ```suggestion
* `-DCMAKE_CXX_FLAGS_DEBUG="-g -fstandalone-debug"` to have full debug information.
```

> during the sycl-toolchain build, after building the complier, it is used to build the device runtime for spirv and nvptx targets. building anything with a debug compiler is very slow, but it can be speedup using

Reviewer: Not super clear sentence. What is the articulation which what precedes and follows?

```bash
export LD_LIBRARY_PATH=$SYCL_HOME/build-Release/lib:$LD_LIBRARY_PATH
```

> this will make the debug compiler select the release libraries. and speedup the build. This only works because the Release and Debug build have ABI compatible configuration, changing LLVM_ENABLE_ASSERTIONS or other configuration may change that.

Reviewer: Insert the previous `export` in the middle of the paragraph here and articulate.

Author: i tried to clarify. this is more of an explaination of the previous export, so I don't think it would fit to insert it in the middle

### Debugging

> using a debug version of XRT as explained above is very useful.

Reviewer: Errr... Yes. And?...

#### Environnement variables

> some environnement variables are very useful for debugging.

Reviewer: ```suggestion

some environment variables are very useful for debugging:

```
Configure your editor to use a spell-checker, or at list configure it for English instead of French ;-)

```bash

> # redirect the directory used as temporary directory 

Reviewer: ```suggestion
# Redirect the directory used as temporary directory 
```

# for the compiler and various tools
export TMP=$SYCL_HOME/../tmp

> # SYCL_PI_TRACE should always be at least at 1, this make the sycl runtime emit logs about which device is selected.

Reviewer: ```suggestion
# SYCL_PI_TRACE should always be at least at 1, this make the SYCL runtime emit logs about which device is selected
```

export SYCL_PI_TRACE=1

> # SYCL_PI_TRACE can be set at -1 to have full debug information but this is quite noisy.

Reviewer: ```suggestion
# SYCL_PI_TRACE can be set at -1 to have full debug information but this is quite noisy
```

export SYCL_PI_TRACE=-1
```

#### Clang flags

> some useful clang flags:

Reviewer: ```suggestion

Some useful Clang flags:
```

> * ```-ccc-print-phases``` outputs the graph of compilation phases.

Reviewer: ```suggestion
* `-ccc-print-phases` outputs the graph of compilation phases;
```

> * ```-ccc-print-bindings``` outputs the mapping from compilation phases to commands.

Reviewer: ```suggestion
* `-ccc-print-bindings` outputs the mapping from compilation phases to commands.
```

> * ```-###``` outputs the commands to run to compile. the driver will create files to reserve them for those commands.

Reviewer: ```suggestion
* `-###` outputs the commands to run to compile. The driver will create files to reserve them for those commands.
```

#### Running a single test

> to run test of the test suite in isolation

Reviewer: ```suggestion

To run test of the test suite in isolation use:
```

```bash
python3 /path/to/build/dir/bin/llvm-lit -v --param XOCC=all path/to/test.cpp
```

> all tests utilities must have been build for this to work.

Reviewer: ```suggestion
where all tests utilities must have been build for this to work.
```

#### v++ Logs

> this kind of error is a beckend issue.

Reviewer: ```suggestion
The kinds of following errors are typical of a back-end issue:
```

```
ERROR: [v++ 60-300] Failed to build kernel(ip) kernel_name, see log for details: ROOT/_x/sycl-xocc.ppArYN_kernel_name/kernel_name/vitis_hls.log
ERROR: [v++ 60-599] Kernel compilation failed to complete
ERROR: [v++ 60-592] Failed to finish compilation
```

> the path contains a hash so they can't be copy pasted.

Reviewer: I do not understand what you mean here.

Author: i meant that path will be different for everyone reading this because they contain a unique hash.

> sometimes the specified log file contains useful information

Reviewer: ```suggestion
Sometimes the specified log file contains useful information
```

> but there is a more detailed log file ```ROOT/_x/sycl-xocc.ppArYN_kernel_name/kernel_name/kernel_name/solution/.autopilot/db/autopilot.flow.log```

Reviewer: ```suggestion
but the more detailed log file `ROOT/_x/sycl-xocc.ppArYN_kernel_name/kernel_name/kernel_name/solution/.autopilot/db/autopilot.flow.log`
```

> in this log we can find the specific command v++ was running when it crashed.

Reviewer: ```suggestion
Then in this log we can find the specific command `v++` was running when it crashed.
```

#### llvm-reduce

> it is possible to use llvm-reduce to track down v++ issues.

Reviewer: ```suggestion
It is possible to use `llvm-reduce` to track down `v++` issues.
```

> first build llvm-reduce

Reviewer: ```suggestion
First build `llvm-reduce` with:
```

```bash

> 

Reviewer: ```suggestion
```

ninja -C build-Release llvm-reduce
```

> than build a bash script called later is_interesting_llvm.sh to exibit the v++ bug.

Reviewer: ```suggestion
then build a script called later `is_interesting_llvm.sh` to exhibit the `v++` bug.
```

> this will look like the following

Reviewer: ```suggestion
This will look like the following:
```

```bash
#!/bin/bash

> # downgrade the IR generate by llvm-reduce

Reviewer: ```suggestion
# Downgrade the IR generate by llvm-reduce
```

./build-Release/bin/opt -verify -S -xoccIRDowngrader $1 -o $1.tmp.ll || exit 1

> # assembler the IR using Vitis's assembler.

Reviewer: ```suggestion
# Assemble the IR using Vitis's assembler
```

.../clang-3.9-csynth/bin/llvm-as $1.tmp.ll -o $1.tmp.xpirbc

> # run the command that crashed directly. on the file generated by llvm-reduce.
> # while grabbing the output

Reviewer: ```suggestion
# Run the command that crashed directly on the file generated by llvm-reduce
# while grabbing the output
```
but I am unsure about the global meaning...

.../clang-3.9-csynth/bin/clang ... -x ir $1.tmp.xpirbc -o - &> out

> # save the exit code of the command. since it should have crashed.

Reviewer: ```suggestion
# Save the exit code of the command since it might have crashed
```

res=$?

> # test that the output matches the original error

Reviewer: ```suggestion
# Test that the output matches the original error
```

cat out | grep "..." > /dev/null || exit 1

> # test the error code

Reviewer: ```suggestion
# Test the error code
```

if [ $res -eq 254 ]; then
# exit 0 to indicate this still exhibit the original bug
  exit 0
else
# exit 1 to indicate this doesn't exhibit the original bug
  exit 1
fi
```

> than to run llvm-reduce.

Reviewer: ?

```bash

> # disassemble the file that crashed Vitis's clang

Reviewer: ```suggestion
# Disassemble the file that crashed Vitis's clang
```

./build-Release/bin/opt -S file_that_crashed_vitis_clang.bc -o tmp.ll

> #run llvm-reduce

Reviewer: ```suggestion
# Run llvm-reduce
```

./build-Release/bin/llvm-reduce --test=./is_interesting_llvm.sh tmp.ll
```

> the reduced IR crashing v++ will be in a file named reduced.ll

Reviewer: ```suggestion
The reduced IR crashing `v++` will be in a file named `reduced.ll`.
```
